srcbpatch/streamreplacer.cpp

```diff
@@ -0,0 +1,544 @@
+        {
+            pNext_->DoReplacements(src_[i], false);
+        }
+        cachedAmount_ = 0;
+        pNext_->DoReplacements(toProcess, true);
+        return;
+    }
+
+    if (src_[cachedAmount_] == toProcess) // check for match
+    {
+        if (++cachedAmount_ >= src_.size())
+        {// send target - do replacement
+            for (size_t q = 0; q < trg_.size(); ++q) { pNext_->DoReplacements(trg_[q], false); }
+            cachedAmount_ = 0;
+        }
+        return;
+    }
+
+    // here:   toProcess is not our char
+    //         lets check for fast track (255/256 probability)
+    if (0 == cachedAmount_)
+    {
+        pNext_->DoReplacements(toProcess, false);
+        return;
+    }
+
+    // here:   We have some cached data
+    //         at least 1 char need to be send further
+    //         remaining cached data including toProcess need to be reprocessed for match
+
+    memcpy(buffer_.data(), src_.data(), cachedAmount_);
+    buffer_[cachedAmount_++]= toProcess;
+    size_t i = 0;
+    do
+    {
+        pNext_->DoReplacements(buffer_[i++], false); // send 1 byte after another
+    } while (0 != memcmp(src_.data(), &buffer_[i], --cachedAmount_));
+    // Everything that was needed has already been sent
+    // cachedAmount_ is zero or greater
+}
+
+
+/// <summary>
+///  creates replacer
+/// </summary>
+/// <param name="src">what we are going to replace</param>
+/// <param name="trg">this is the result of the replacement</param>
+/// <returns>Replacer for building replacement chain</returns>
+static unique_ptr<StreamReplacer> CreateSimpleReplacer(
+    unique_ptr<AbstractBinaryLexeme>& src,  // what to replace
+    unique_ptr<AbstractBinaryLexeme>& trg)  // with what
+{
+    return unique_ptr<StreamReplacer>(new UsualReplacer(src, trg));
+}
+
+
+//--------------------------------------------------
+///
+///      |--SRC 1  TRG 1  |
+///  O - |-- ...          | - o
+///      |--SRC N  TRG N  |
+/// 
+class ChoiceReplacer final : public ReplacerWithNext
+{
+    typedef struct
+    {
+        span<const char> src_;
+        span<const char> trg_;
+    }ChoiceReplacerPair;
+
+public:
+    /// <summary>
+    ///   creating ChoiceReplacer from provided pairs
+    /// </summary>
+    /// <param name="choice">vector os source & target pairs</param>
+    ChoiceReplacer(StreamReplacerChoice& choice)
+    {
+        size_t bufferSize = 0; // to allocate buffer
+        const size_t sz = choice.size();
+        rpairs_.resize(sz);
+        for (size_t i = 0; i < sz; ++i)
+        {
+            auto& vPair = choice[i]; // copy from
+            auto& rpair = rpairs_[i];// copy to
+
+            rpair.src_ = vPair.first->access();
+            const size_t sourceSize = rpair.src_.size();
+            if (bufferSize < sourceSize)
+            {
+                bufferSize = sourceSize; // calculate necessary buffer size
+            }
+
+            rpair.trg_ = vPair.second->access();
```

Comment:
Don't economize letters ... target is not such big word. 
vPair, rpair - the same thing. 
rpair - replacePair (I can guess it)
vPair - I have no idea what it is. I have associations with the vTable, where 'v' stands for virtual.

Reply:
v - variable

and yes - it's hard to find good name.  at least - src exists inside of the pair =)

Reply:
vPair  -> fromToLexemePair ?